File: MakeChange.rb
def make_change(target,coins,used_coins=[])
  unless target < 1
    rankings = coins.map do |coin|
      coin > target ? (coin * target) : (target % coin) + (target/coin) * 2.71828
    end

    best_coin_index = rankings.index(rankings.min)
    used_coins<<coins[best_coin_index]
    target -= coins[best_coin_index]

    make_change(target,coins,used_coins)
  else
    if target == 0
      used_coins
    else
      nil
    end
  end
end

File: MakeChange.js
var makeChange = function(coins, amount){
  var ranks = [];

  for (var i = 0; i < coins.length; i++) {
    var coin = coins[i];
    if (coin > amount) {
      ranks.push(1000000.0);
    } else {
      ranks.push((amount % coin) + (Math.round(amount / coin) - 1) * 2.71828);
    }
  }

  var bestCoin = coins[ranks.indexOf(min(ranks))];

  if (amount - bestCoin === 0) {
    return [bestCoin];
  } else if (amount > 0) {
    return [bestCoin].concat(makeChange(coins, (amount - bestCoin)));
  }
};